x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java

```diff
@@ -501,9 +536,11 @@ public void onFailure(Exception e) {
     }
 
     private void putComposableTemplate(
+        ClusterState state,
         final String templateName,
         final ComposableIndexTemplate indexTemplate,
-        final AtomicBoolean creationCheck
+        final AtomicBoolean creationCheck,
+        final boolean isUpgrade
```

Comment:
This is not a change related to this PR, it is how this registry is handling concurrency across all component upgrades.

> do we need to keep the flag true for the entire execution of the rollover

I believe we do, otherwise we may get into concurrent rollover request for the same data stream. I don't know what the exact consequences of that, but definitely sounds like something we better avoid. What's the down side? If a composable template gets upgraded and we find data streams that are affected by it, why should we allow for additional upgrade **for the same composable template** before rollovers are done?

Eventually, our hope is to have this as a basic Elasticsearch capability (like proposed in  https://github.com/elastic/elasticsearch/issues/96521#issuecomment-1695801781) that the registries only opt-in for. I assume that any such implementation will use the same atomicity semantics, where the composable template upgrade action includes the subsequent rollovers as part of it before it can be considered done.

Reply:
> I believe we do, otherwise we may get into concurrent rollover requests for the same data stream.
How might we end up in that scenario? This would require the composable template to be changed and for the template installation process to kick off again correct?

Again, this comment is mostly seeing if we can improve the structure of the code. If that comes at a cost to the correctness of the behavior then we can ignore the suggestion.

Reply:
This is mainly where I need your input from the review - is the code **correct**? 
My assumption is that we keep the former behavior of avoiding concurrent index template upgrades, and that rollovers of affected data streams should be considered now part of the index template upgrade. Is this assumption correct? Is the code properly enforcing that?

Reply:
This code does seem to be correctly enforcing that. My comments here were just questioning if waiting for rollovers to complete before marking the template install complete was something that made sense to do or not. For now, I think it's fine to have that be the case.